src/jpgimage.cpp

```diff
@@ -706,19 +707,22 @@ namespace Exiv2 {
                         }
                     } else if (bPrint) {
                         out << "| " << Internal::binaryToString(buf, size > 32 ? 32 : size, size > 0 ? 2 : 0);
-                        if (std::strcmp(signature, iccId_) == 0) {
-                            int chunk = (int)signature[12];
-                            int chunks = (int)signature[13];
+                        if (signature.compare(iccId_) == 0) {
+                            // construct a new string from the whole DataBuf, as have to read beyond the null
+                            // termination which marks the end of the string signature
+                            const std::string full_signature((const char*)buf.pData_ + 2, buf.size_ - 2);
```

Comment:
There is something I do not understand here. Was the code overflowing before in this block? Was is happening only for some bad formed images?

Would not be enough to check if the buf.size_ is big enough to read those bytes without constructing another `std::string` ?

Reply:
The code is tricky. It read a certain amount of data from the image, which may or may not contain a `0` and then treats the thing as C string. However, in most cases there is a null terminator inside the read data, e.g.:
```
ICC_PROFILE\0SomeNumbers
```
This would have been the C string `signature`. Now, since all C functions assume null terminated strings, you can `strcpm` this thing with `"ICC_PROFILE"` and you'll get `0`, as C does not know that there are meaningful data beyond the terminator. (Apparently the next two numbers after `\0` are the chunk numbers.) That does however not work with a C++ string, if you construct it via the constructor where you specify the length (as that one permits a `0` inside the string).

Since I changed this thing to a `std::string` but still wanted to be able to compare it, I had to make `signature` stop at the null terminator. To be able to read beyond the `\0` requires another `std::string` which spans the whole `DataBuf` (actually it can be done differently, as you mentioned, I'll probably do that as the string is not really required and I am not sure if compilers are smart enough to optimize that away as `std::string` will probably use `malloc`).

Reply:
I changed the code to directly access the array & tried to improve the comment to explain matters better.